app/src/org/commcare/activities/GlobalPrivilegeClaimingActivity.java

```diff
@@ -101,72 +105,45 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         switch(requestCode) {
             case BARCODE_CAPTURE:
                 if (resultCode == RESULT_OK) {
-                    String[] fields = processScanResult(data.getStringExtra("SCAN_RESULT"));
-                    if (fields == null) {
-                        privilegeClaimFailed();
-                    } else {
-                        String flag = fields[0];
-                        String username = fields[1];
-                        String signature = fields[2];
-                        if (checkProperFormAndAuthenticity(flag, username, signature)) {
-                            GlobalPrivilegesManager.enablePrivilege(flag, username);
-                            refreshUI();
-                        } else {
-                            privilegeClaimFailed();
+                    String scanResult = data.getStringExtra("SCAN_RESULT");
+                    try {
+                        Pair<String, String[]> activatedPrivileges =
+                                new PrivilegesUtility(GlobalConstants.TRUSTED_SOURCE_PUBLIC_KEY).
+                                        processPrivilegePayloadForActivatedPrivileges(scanResult);
+
+                        for(String p : activatedPrivileges.second) {
+                            if (!GlobalPrivilegesManager.allGlobalPrivilegesList.contains(p)) {
+                                Log.d(TAG, "Request to activate unknown privilege: " + p);
+                            } else {
+                                GlobalPrivilegesManager.enablePrivilege(p, activatedPrivileges.first);
+                            }
                         }
+                        refreshUI();
+
+                    } catch (PrivilegesUtility.UnrecognizedPayloadVersionException e) {
+                        e.printStackTrace();
+                        privilegeClaimTooNew();
+                    } catch (PrivilegesUtility.PrivilagePayloadException e) {
+                        e.printStackTrace();
+                        privilegeClaimFailed();
                     }
                 }
         }
     }
 
-    private void privilegeClaimFailed() {
+    private void privilegeClaimTooNew() {
```

Comment:
Can you call this something like `privilegePayloadVersionTooNew()` instead? `privilegeClaimTooNew` sounds more to me like the situation where the device is trying to claim a privilege that it doesn't know exists yet